beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go

```diff
@@ -197,86 +194,146 @@ func (vs *Server) BuildBlockParallel(ctx context.Context, sBlk interfaces.Signed
+	}
+	if vs.BlockBuilder == nil || !vs.BlockBuilder.Configured() {
+		return nil, nil, errors.New("unconfigured block builder")
+	}
+
+	copiedBlock, err := vs.copyAndCastBlock(block)
 	if err != nil {
-		return nil, errors.Wrap(err, "could not unblind builder block")
+		return nil, nil, err
 	}
 
-	// Broadcast the new block to the network.
-	blkPb, err := blk.Proto()
+	payload, bundle, err := vs.BlockBuilder.SubmitBlindedBlock(ctx, block)
 	if err != nil {
-		return nil, errors.Wrap(err, "could not get protobuf block")
+		return nil, nil, errors.Wrap(err, "submit blinded block failed")
 	}
-	if err := vs.P2P.Broadcast(ctx, blkPb); err != nil {
-		return nil, fmt.Errorf("could not broadcast block: %v", err)
+
+	if err := copiedBlock.UnBlind(payload); err != nil {
+		return nil, nil, errors.Wrap(err, "unblind failed")
 	}
 
-	root, err := blk.Block().HashTreeRoot()
+	sidecars, err := unblindBlobsSidecars(copiedBlock, bundle)
 	if err != nil {
-		return nil, fmt.Errorf("could not tree hash block: %v", err)
+		return nil, nil, errors.Wrap(err, "unblind sidecars failed")
 	}
-	log.WithFields(logrus.Fields{
-		"blockRoot": hex.EncodeToString(root[:]),
-	}).Debug("Broadcasting block")
-
-	if blk.Version() >= version.Deneb {
-		if !blinded {
-			dbBlockContents := req.GetDeneb()
-			if dbBlockContents == nil {
-				return nil, errors.New("signed beacon block contents is empty")
-			}
-			scs, err = buildBlobSidecars(blk, dbBlockContents.Blobs, dbBlockContents.KzgProofs)
-			if err != nil {
-				return nil, fmt.Errorf("could not build blob sidecars: %v", err)
-			}
-		}
-		for i, sc := range scs {
-			if err := vs.P2P.BroadcastBlob(ctx, uint64(i), sc); err != nil {
-				log.WithError(err).Error("Could not broadcast blob")
-			}
-			readOnlySc, err := blocks.NewROBlobWithRoot(sc, root)
-			if err != nil {
-				return nil, fmt.Errorf("could not create ROBlob: %v", err)
-			}
-			verifiedSc := blocks.NewVerifiedROBlob(readOnlySc)
-			if err := vs.BlobReceiver.ReceiveBlob(ctx, verifiedSc); err != nil {
-				log.WithError(err).Error("Could not receive blob")
-			}
-		}
+
+	return copiedBlock, sidecars, nil
+}
+
+// copyAndCastBlock creates a copy of a SignedBeaconBlock.
+func (vs *Server) copyAndCastBlock(block interfaces.SignedBeaconBlock) (interfaces.SignedBeaconBlock, error) {
+	copied, err := block.Copy()
+	if err != nil {
+		return nil, errors.Wrap(err, "copy failed")
+	}
+	castedBlock, ok := copied.(interfaces.SignedBeaconBlock)
+	if !ok {
+		return nil, errors.New("cast to SignedBeaconBlock failed")
 	}
+	return castedBlock, nil
+}
 
-	if err := vs.BlockReceiver.ReceiveBlock(ctx, blk, root); err != nil {
-		return nil, fmt.Errorf("could not process beacon block: %v", err)
+// handleUnblindedBlock processes unblinded beacon blocks.
+func (vs *Server) handleUnblindedBlock(block interfaces.SignedBeaconBlock, req *ethpb.GenericSignedBeaconBlock) ([]*ethpb.BlobSidecar, error) {
+	dbBlockContents := req.GetDeneb()
```

Comment:
Checking `if block.Version() < version.Deneb` is probably better

Reply:
there's no version here, `ethpb.GenericSignedBeaconBlock` returns the actual protobuf type